### PR TITLE
Change nox backend to `uv`

### LINKED ISSUE #286: Add `uv` support for CI
### Feature description

[uv](https://github.com/astral-sh/uv) is a new package manager for Python, built in Rust by astral-sh. Large [improvements](https://github.com/wntrblm/nox/pull/762) have been shown over standard `pip`, while maintaining `pip` functionality (`uv pip install xx`). This issue is to investigate adding `uv` to our nox configuration for a reduction in test suite runtimes. 

For reference, here is the PyBaMM issue for adding `uv` support: https://github.com/pybamm-team/PyBaMM/issues/3825

### Motivation

_No response_

### Possible implementation

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by BradyPlanden**
# Description

Change nox backend to `uv` and updates nox with optional dependency.

## Issue reference
Fixes #286 

## Review
Before you mark your PR as ready for review, please ensure that you've considered the following:
- Updated the [CHANGELOG.md](https://github.com/pybop-team/PyBOP/blob/develop/CHANGELOG.md) in reverse chronological order (newest at the top) with a concise description of the changes, including the PR number.
- Noted any breaking changes, including details on how it might impact existing functionality.

## Type of change
- [ ] New Feature: A non-breaking change that adds new functionality.
- [ ] Optimization: A code change that improves performance.
- [ ] Examples: A change to existing or additional examples.
- [ ] Bug Fix: A non-breaking change that addresses an issue.
- [ ] Documentation: Updates to documentation or new documentation for new features.
- [ ] Refactoring: Non-functional changes that improve the codebase.
- [ ] Style: Non-functional changes related to code style (formatting, naming, etc).
- [X] Testing: Additional tests to improve coverage or confirm functionality.
- [ ] Other: (Insert description of change)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybop-team/PyBOP/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All unit tests pass: `$ nox -s tests`
- [ ] The documentation builds: `$ nox -s doctest`

You can run integration tests, unit tests, and doctests together at once, using `$ nox -s quick`.

## Further checks:
- [ ] Code is well-commented, especially in complex or unclear areas.
- [ ] Added tests that prove my fix is effective or that my feature works.
- [ ] Checked that coverage remains or improves, and added tests if necessary to maintain or increase coverage.

Thank you for contributing to our project! Your efforts help us to deliver great software.

### CODE REVIEW

File: noxfile.py
Comment (agriyakhetarpal) [inline noxfile.py:8]: I know this isn't ready for review but it looks complete to me. I think adding this minimum version would help developers receive a graceful error message and let them upgrade their `nox` installations. This is the second release that supports `uv` and has some relevant bug fixes.

```suggestion
nox.options.force_venv_backend = "uv"
nox.needs_version  = ">=2024.4.15"

```

**Review by agriyakhetarpal — APPROVED** [review]

File: noxfile.py
Comment (BradyPlanden) [reply noxfile.py:8]: Thanks @agriyakhetarpal - yes this is probably good to review. I'm still on the fence on whether we move it forward, although it appears to be robust across the CI. I didn't see a massive increase in performance over the nox backend with pip unforunately (~1min I believe).  If we run into issues from users, it's an easy revert though. 

Thanks for the review!

Comment (agriyakhetarpal) [reply noxfile.py:8]: Waiting for a while sounds good to me. I think it's much more useful for local development where you might have cached wheels present, and for those packages where you have several required dependencies (maybe more than a hundred). If all of the dependencies sre available as wheels (which they are), it should be okay. On the PyBaMM side we haven't implemented this because we are waiting for a few big projects to do it, so I'll circle back here when that happens in case you decide to hold off until that happens, but you can feel free to merge maybe after the next release.

**Comment by BradyPlanden** [comment]
I've decided to merge this, as the local testing suite time is becoming combersome. If anyone finds issues with `uv` and is looking at this PR, please open an issue.

**Pull request merged**
